### PR TITLE
Makes cloud-init execution optional

### PR DESCRIPTION
**Pull request by mpagot**
Add setting to optionally disable cloud-init execution during the deployment. Az cloud-init is mostly in charge to configure the web server, if disabled the installation and configuration is executed later and by some ssh commands.

- Related ticket: https://jira.suse.com/browse/TEAM-9428

# Verification run: 

### CLOUDINIT
test with setting IPADDR2_CLOUDINIT=1
- http://openqaworker15.qa.suse.cz/tests/290913 :red_circle:  exotic ip2t-snet issue ??
- http://openqaworker15.qa.suse.cz/tests/290923 :red_circle: `[{"code":"InUseSubnetCannotBeUpdated","message":"Subnet ip2t290922-snet is in use and cannot be updated."`
- http://openqaworker15.qa.suse.cz/tests/290934 :green_circle:  it is using cloud-init (settings works), cloud-init script content is fine http://openqaworker15.qa.suse.cz/tests/290934/logfile?filename=deploy-cloud-init-web.txt and available as job attachment, `deployment` and `configure` stages are fine. It fails in `sanity` when `systemctl is-system-running` return `degradated`.

### NOCLOUDINIT
test with setting IPADDR2_CLOUDINIT=0
sle-15-SP5-SapCloud-Azure-Payg-x86_64-BuildLATEST_AZURE_SLE15_5_PAYG-ipaddr2_azure_test@64bit
- http://openqaworker15.qa.suse.cz/tests/290817 :red_circle: fails in az NIC 
- http://openqaworker15.qa.suse.cz/tests/290912 :red_circle:  exotic ip2t-snet issue ??
- http://openqaworker15.qa.suse.cz/tests/290922 :red_circle:  ip2t-snet issue
- http://openqaworker15.qa.suse.cz/tests/290935 :red_circle:  ip2t-snet issue

### CLOUDINIT
test with setting IPADDR2_CLOUDINIT=1

sle-15-SP5-SapCloud-Azure-Byos-x86_64-BuildLATEST_AZURE_SLE15_5_BYOS-ipaddr2_azure_test@64bit with SCC code
-  http://openqaworker15.qa.suse.cz/tests/290906 :red_circle: fails in `az run-command` 
- http://openqaworker15.qa.suse.cz/tests/290919 :green_circle: it is using cloud-init and content is fine http://openqaworker15.qa.suse.cz/tests/290919/logfile?filename=deploy-cloud-init-web.txt, it has reg code in cloudinit script. `deployment` and `configure` are fine. Fails in `sanity`

test without SCC
http://openqaworker15.qa.suse.cz/tests/290815#step/deploy/65 :green_circle: fails as expected
 
### NOCLOUDINIT
test with setting IPADDR2_CLOUDINIT=0

 sle-15-SP5-SapCloud-Azure-Byos-x86_64-BuildLATEST_AZURE_SLE15_5_BYOS-ipaddr2_azure_test@64bit with SCC code
 - http://openqaworker15.qa.suse.cz/tests/290911 :red_circle:  fails in az NIC command 
 - http://openqaworker15.qa.suse.cz/tests/290918 :red_circle: timeout as running registration http://openqaworker15.qa.suse.cz/tests/290918#step/configure/160
 -  http://openqaworker15.qa.suse.cz/tests/290995

### CODE REVIEW

File: lib/sles4sap/ipaddr2.pm [hunk 1]
```diff
@@ -108,8 +135,12 @@ sub ipaddr2_azure_deployment {
         region => $args{region});
 
     # Create a VNET only needed later when creating the VM
-    my $vnet = DEPLOY_PREFIX . '-vnet';
-    my $subnet = DEPLOY_PREFIX . '-snet';
+    # Use $rg instead of DEPLOY_PREFIX to try to prevent
+    # some deployment failures like:
+    #    Subnet(ip2t-snet) does not exist, but failed to create a new subnet with address prefix 10.0.0.0/24. It may be caused by name or address prefix conflict. Please specify an appropriate subnet name with --subnet or a valid address prefix value with --subnet-address-prefix.[0m
```
Comment (alvarocarvajald) [inline lib/sles4sap/ipaddr2.pm:140]: Perhaps this long comment can be broken in several lines.

File: lib/sles4sap/ipaddr2.pm [hunk 2]
```diff
@@ -166,22 +197,47 @@ sub ipaddr2_azure_deployment {
...
+        my $cloud_init_content = <<END;
+#cloud-config
+package_upgrade: false${register_str}
+packages:
+  - nginx
+runcmd:
+  - 'echo "I am \$(hostname)" > /srv/www/htdocs/index.html'
+  - sudo systemctl enable nginx.service
+  - sudo systemctl start nginx.service
+END
```
Comment (alvarocarvajald) [inline lib/sles4sap/ipaddr2.pm:222]: I think the way this is constructed is a bit confusing. Why not something like:
```
if ($args{cloudinit}) {
    my $cloud_init_content = <<END;
#cloud-config
package_upgrade: false
packages:
  - nginx
runcmd:
  - 'echo "I am \$(hostname)" > /srv/www/htdocs/index.html'
  - sudo systemctl enable --now nginx.service

END

    if ($args{scc_code}) {
        $cloud_init_content .= <<END
bootcmd:
  - registercloudguest --clean
  - registercloudguest --force-new -r $args{scc_code}
END
    }
}
```
After all, it looks like cloud-init config is YAML and `bootcmd` is not a sub-element of anything else. Or am I misunderstanding?

File: lib/sles4sap/ipaddr2.pm [hunk 3]
```diff
@@ -166,22 +197,47 @@ sub ipaddr2_azure_deployment {
...
+        $cloud_init_file = '/tmp/cloud-init-web.txt';
+        write_sut_file($cloud_init_file, $cloud_init_content);
+        # uplaod to allow debugging
```
Comment (alvarocarvajald) [inline lib/sles4sap/ipaddr2.pm:225]: s/uplaod/upload/

File: lib/sles4sap/ipaddr2.pm [hunk 4]
```diff
@@ -999,6 +1060,123 @@ sub ipaddr2_create_cluster {
...
+        bastion_ip => $args{bastion_ip});
+    ipaddr2_ssh_internal(id => $args{id},
+        cmd => 'echo "I am $(hostname)" > /srv/www/htdocs/index.html',
+        bastion_ip => $args{bastion_ip});
+    ipaddr2_ssh_internal(id => $args{id},
+        cmd => 'sudo systemctl enable nginx.service',
+        bastion_ip => $args{bastion_ip});
+    ipaddr2_ssh_internal(id => $args{id},
+        cmd => 'sudo systemctl start nginx.service',
+        bastion_ip => $args{bastion_ip});
```
Comment (alvarocarvajald) [inline lib/sles4sap/ipaddr2.pm:1177]:
```suggestion
    my @cmds = ('sudo zypper install -y nginx', 'echo "I am $(hostname)" > /srv/www/htdocs/index.html', 'sudo systemctl enable --now nginx.service');
    ipaddr2_ssh_internal(id => $args{id}, cmd => $_, bastion_ip => $args{bastion_ip}) for (@cmds);
```

Comment (alvarocarvajald) [reply lib/sles4sap/ipaddr2.pm:1177]: Or:
```
ipaddr2_ssh_internal(id => $args{id}, cmd => $_, bastion_ip => $args{bastion_ip})
    for ('sudo zypper install -y nginx', 'echo "I am $(hostname)" > /srv/www/htdocs/index.html', 'sudo systemctl enable --now nginx.service');
```
But not sure how tidy will react to that ;)

File: data/sles4sap/cloud-init-web.txt
Comment (mpagot) [inline data/sles4sap/cloud-init-web.txt:1]: deleted as now it is composed on the fly

File: lib/sles4sap/ipaddr2.pm (hunk 2 above)
Comment (mpagot) [reply lib/sles4sap/ipaddr2.pm:222]: I'm not sure if the order matter or not, but in any case I'll give it a try to your suggestion as code is much cleaner

**Review by lpalovsky — APPROVED** [review]
LGTM aside the weird check failure

**Comment by badboywj** [comment]
LGTM, thanks

**Review by alvarocarvajald — APPROVED** [review]
LGTM

**Comment by alvarocarvajald** [comment]
> LGTM aside the weird check failure

Seems registry.opensuse.org was not available while running the isotovideo check yesterday. It's responding now, so I restarted the job.

**Pull request merged**